### DISCUSSION TITLE
Route53 alias entry in apex domains

### DISCUSSION THREAD

**Discussion by fsanzdev** [gruntwork-io/knowledge-base#781] [Help] [answered]
Hi, 

I want to create an entry in Route53 which is an alias to a Cloudfront distribution, but it looks like the module (networking/route53) doesn't support it. Could you confirm this and if there is a workaround?

Thanks.

---

<ins datetime="2023-11-06T13:28:05Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110572">Tracked in ticket #110572</a></p>
</ins>

**Comment by gitsstewart** [comment]
@arsci can you confirm?

**Comment by fsanzdev** [comment]
Anything regarding this? It's quite limiting not being able to define alias for apex domain @arsci @gitsstewart

**Comment by arsci** [comment]
Hi @fsanzdev, are you able to create an alias record by adding an alias block within the subdomain block as the variables passed into the module? I tested this today in my own account, and I was able to a route53 alias A record pointed to a CloudFront distribution. [This link](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/networking/route53/variables.tf#L132:L144) should go directly to the module variable input comments on how to structure this input to create an alias record.

Alternatively, if you used something like the `s3-cloudfront` module from `terraform-aws-static-assets`, there are configurations that allow the Route53 entries to be created with your CloudFront distribution. 

Let me know if this helps! I'm happy to share the code I used to test the `networking/route53` module if that would be helpful.

**Comment by fsanzdev** [comment]
Hi @arsci, thanks for your reply. Yes, in subdomains it's available, but as I mentioned in the title the issue is in the apex domain section. It looks like it's not possible to set aliases in apex domains which limits the configuration options for Cloudfront in this case, but for any kind of aliases in any other cases.

**Reply by arsci** [reply]
Hi @fsanzdev, you should be able to use a [similar alias block](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/networking/route53/variables.tf#L171:L183) in the `apex_records` block as well. Note that the structure is a little different from the `alias` block when used in the `subdomain` block.

I successfully tested this with my own domain today (domain values changed in this snippet) against a CloudFront distribution I previously created in my own AWS account.

```
public_zones = {
  "mydomain.com" = {
    created_outside_terraform = true
    hosted_zone_domain_name = "mydomain.com"
    subject_alternative_names = []
    apex_records = [
      {
        type = "A"
        records = null
        ttl = null
        alias = {
          name = "distribution_id.cloudfront.net"
          evaluate_target_health = false
          zone_id = "Z2FDTNDATAQYW2" ## CloudFront Hosted Zone ID
        }
      }
    ]
    tags = {}
  }
}
```

**Reply by fsanzdev** [reply]
I've tried using your sample code and it didn't work (using version 107.6 from catalog-service). This is my code:

```
,
        {
        type = "A"
        records = null
        ttl = null
        alias = {
          name = "xxxxxxxxxx.cloudfront.net."
          evaluate_target_health = false
          zone_id = "XXXXXXXXXXXXX" ## CloudFront Hosted Zone ID
        }
      }
```

And the error I get is:

```
 Error: Unsupported attribute
│ 
│   on main.tf line 316, in locals:
│  316:           alias   = record.alias
│ 
│ This object does not have an attribute named "alias".
╵
╷
│ Error: Unsupported attribute
│ 
│   on main.tf line 316, in locals:
│  316:           alias   = record.alias
│ 
│ This object does not have an attribute named "alias".
╵
╷
│ Error: Unsupported attribute
│ 
│   on main.tf line 316, in locals:
│  316:           alias   = record.alias
│ 
│ This object does not have an attribute named "alias".

```

**Reply by fsanzdev** [reply, marked as answer]
Ok, we just found the issue. We had several entries in the apex record (only one of them had an alias) but all of them need the entry for an alias even if it's empty (alias={}) 

I think this behaviour should be documented somewhere in the variables.tf file where there are several samples and it explicitly says `# (Optional) An alias block.` when referring to the alias entry.
